lib/einhorn/command.rb

```diff
@@ -116,39 +116,65 @@ def self.signal_all(signal, children=nil, record=true)
         rescue Errno::ESRCH
+              Einhorn.log_debug("Attempted to #{signal} child #{child.inspect} but the process does not exist", :upgrade)
         else
           signaled[child] = spec
         end
       end
 
       if Einhorn::State.signal_timeout && record
         Einhorn::Event::Timer.open(Einhorn::State.signal_timeout) do
-          children.each do |child|
-            spec = Einhorn::State.children[child]
-            next unless spec # Process is already dead and removed by mourn
-            signaled_spec = signaled[child]
-            next unless signaled_spec # We got ESRCH when trying to signal
-            if spec[:spinup_time] != signaled_spec[:spinup_time]
-              Einhorn.log_info("Different spinup time recorded for #{child} after #{Einhorn::State.signal_timeout}s. This probably indicates a PID rollover.", :upgrade)
-              next
-            end
-
-            Einhorn.log_info("Child #{child.inspect} is still active after #{Einhorn::State.signal_timeout}s. Sending SIGKILL.")
-            begin
-              Process.kill('KILL', child)
-            rescue Errno::ESRCH
+          if Einhorn::State.config[:resignal_timeout]
+            resignal_children
+          else
+            children.each do |child|
+              spec = Einhorn::State.children[child]
+              next unless spec # Process is already dead and removed by mourn
+              signaled_spec = signaled[child]
+              next unless signaled_spec # We got ESRCH when trying to signal
+              if spec[:spinup_time] != signaled_spec[:spinup_time]
+                Einhorn.log_info("Different spinup time recorded for #{child} after #{Einhorn::State.signal_timeout}s. This probably indicates a PID rollover.", :upgrade)
+                next
+              end
+
+              Einhorn.log_info("Child #{child.inspect} is still active after #{Einhorn::State.signal_timeout}s. Sending SIGKILL.")
+              begin
+                Process.kill('KILL', child)
+              rescue Errno::ESRCH
+              end
+              spec[:signaled].add('KILL')
             end
-            spec[:signaled].add('KILL')
           end
+
+          "Successfully sent #{signal}s to #{signaled.length} processes: #{signaled.keys}"
         end
       end
+    end
 
-      "Successfully sent #{signal}s to #{signaled.length} processes: #{signaled.keys}"
+    def self.resignal_children
+      Einhorn::State.children.select{|_, c| c[:signaled] && c[:signaled].length > 0}.each do |pid, child|
+        next unless child[:last_signaled_at]
+
+        now = Time.now
+        expires_at = child[:last_signaled_at] + Einhorn::State.signal_timeout
+        next unless now >= expires_at
+
+        Einhorn.log_info("Child #{pid.inspect} was signaled #{(child[:last_signaled_at] - now).to_i}s ago. Sending SIGKILL as it is still active after #{Einhorn::State.signal_timeout}s timeout.", :upgrade)
+        begin
+          Process.kill('KILL', pid)
```

Comment:
Hm! I'm not sure what the difference is between this and the `Child #{child.inspect} is still active after #{Einhorn::State.signal_timeout}s. Sending SIGKILL` code block. Why would this get hit, and the other wouldn't? (I guess it has to do with the `last_signaled_at` portion, but I thought the code above should still SIGKILL things, when it really doesn't seem to?

Reply:
No problem, the change is subtle and the diff doesn't show it all that great.

The original behaviour in `signal_all` was optimistic and assume the timer would always work. That is, it would schedule a timer that would act only on the local `children` values. If that timer didn't work for one reason or another (as has been observed with "stuck workers"), those processes would never be looked at again as it assumes the timer handled them properly.

The new block is more pessimistic. It will not concern itself with what was just signaled, instead it will look at _all_ processes that have been signaled and compare their `last_signaled_at` timestamp delta with the `signal_timeout` value. This way, the timer in `signal_all` will always try to ensure processes that should be shut down are shutdown.